### DISCUSSION TITLE
DSGVO-proofing mkdocs?

### DISCUSSION THREAD

**Discussion by koffeiniker** [squidfunk/mkdocs-material#4855] [Ideas]
One thing, that bothers me with many static site generators like hugo and mkdocs is the legal traps they place.

Background: following European law like DSGVO / Datenschutzgrundverordnung (GDPR in the english namespace), it is somewhere between highly problematic and illegal to remotely include resources like WebFonts or Javascript. Now mkdocs/material appears to include Google Fonts, while these might also just be included to the image and referenced locally. Locally they pose no threat, remotely they are literally "mating calls" for greedy lawyers to sue you, if you publish an mkdocs site on the internet. 

I can virtually hear you: "this is why we include DSGVO statement". Regrettably these won't save you, because you literally employ Google, rsp. cloudflare as IT-subcontractors (for free, but legally relevant). According to German law, I believe (I am not a lawyer, just reflecting some press reports), there is the right for deleting. Now if I publish an mkdocs site, any user has the right to have their IP addresses deleted from my logs (sounds idiotic, but is just like that) and **even worse, that user also has the right to demand ME to get his IP address deleted from cloudflare's logs, rsp. Google's logs, which usually I will not be able to do**.

I would propose to try to identify the places in the theme, that call up for remote resources and make them local. For a usual SSG, I'd say "hey guy, hack your theme yourself and get into a safe state", but mkdocs in contrast should be like "no servicable parts inside", so here I would expect the thing to be legally uncritical in the first place.

I am open for comments.

**Comment by squidfunk** [comment]
Thanks for asking. We are very aware of the situation. This is _exactly_ the raison d'être for the [built-in privacy plugin](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/), which will not only inline all external assets used by the theme (i.e. those that we manage), but also those that the author might add (i.e. you manage). It's essentially a safety net for the situation you stated.

> I would propose to try to identify the places in the theme, that call up for remote resources and make them local. For a usual SSG, I'd say "hey guy, hack your theme yourself and get into a safe state", but mkdocs in contrast should be like "no servicable parts inside", so here I would expect the thing to be legally uncritical in the first place.

Our documentation precisely explains why this is not possible without losing convenience and flexibility. See the section on [external assets](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/#external-assets), scroll down and find the admonition which says "Why can't Material for MkDocs bundle all assets by design?":

<img width="733" alt="Bildschirmfoto 2023-01-12 um 11 05 00" src="https://user-images.githubusercontent.com/932156/212037786-359a2c21-f640-40e5-bfac-3c64ede6ac8a.png">

**Comment by koffeiniker** [comment]
After checking the plugin you mentioned and learning that this involves a monthly fee, I decided to go another way. I do not earn money with this, at least currently, so I'd prefer to keep my investment in time rather than money. I still think what you do is a good idea and for companies some peanuts for big benefit. My mileage regrettably varies.

So for my own part I will now create a script to scan the resulting site and replace the links with downloads at build-time. I can weave this into a separate stage in the pipeline and it will work with "any" SSG, .i.e. for me with mkdocs and hugo. During development, I would not care.

Consider the matter closed. ;-)

**Reply by squidfunk** [reply]
I'm glad that you found a solution that works for you! Feel free to share it here with other users, if you like.

Funnily, this is exactly what the built-in privacy plugin does. Yes, it's currently reserved to sponsors, because it required significant effort to pull off – remember that Material for MkDocs does cater to __a lot__ of use cases – but I expect it to be released in the community edition in the coming months. After all, sponsorships enabled us to implement the plugin in the first place 😊
